Mathlib/SetTheory/ZFC/Rank.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2024 Dexin Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dexin Zhang
+-/
+import Mathlib.SetTheory.Ordinal.Arithmetic
+import Mathlib.SetTheory.ZFC.Basic
+
+/-!
+# Ordinal ranks of ZFSet
+
+In this file, we define the ordinal ranks of `ZFSet`. This rank is the same as `WellFounded.rank`
+over `∈`, but defined as `Quotient.lift` of `PSet.rank` so that the universe level is
+`ZFSet.{u} → Ordinal.{u}`.
+
+## Definitions
+
+* `PSet.rank`: Ordinal rank of a pre-set.
+* `ZFSet.rank`: Ordinal rank of a ZFC set.
+-/
+
+universe u v
+
+open Ordinal Order
+
+namespace PSet
+
+/-- The ordinal rank of a pre-set -/
+noncomputable def rank : PSet.{u} → Ordinal.{u}
+  | ⟨_, A⟩ => lsub fun a => rank (A a)
+
+theorem rank_eq_of_equiv : (x y : PSet) → Equiv x y → rank x = rank y
+  | ⟨_, _⟩, ⟨_, _⟩, ⟨αβ, βα⟩ =>
+    lsub_eq_of_range_eq (by
+      ext; constructor <;> simp <;> intro a h
+      · obtain ⟨b, h'⟩ := αβ a
+        exists b; rw [← h, rank_eq_of_equiv _ _ h']
+      · obtain ⟨b, h'⟩ := βα a
+        exists b; rw [← h, rank_eq_of_equiv _ _ h'])
+
+theorem rank_lt_of_mem : {x y : PSet} → y ∈ x → rank y < rank x
+  | ⟨_, _⟩, _, ⟨_, h⟩ => by
+    rw [rank_eq_of_equiv _ _ h]
+    apply lt_lsub
+
+theorem rank_le_of_forall_mem_rank_lt {o : Ordinal} :
+    {x : PSet} → (∀ y ∈ x, rank y < o) → rank x ≤ o
+  | ⟨_, A⟩, h => lsub_le fun a => h (A a) (Mem.mk A a)
+
+end PSet
+
+namespace ZFSet
+
+variable {x y : ZFSet.{u}}
+
+/-- The ordinal rank of a ZFC set -/
+noncomputable def rank : ZFSet.{u} → Ordinal.{u} :=
+  Quotient.lift PSet.rank PSet.rank_eq_of_equiv
+
+theorem rank_lt_of_mem : y ∈ x → rank y < rank x :=
+  Quotient.inductionOn₂ x y fun _ _ => PSet.rank_lt_of_mem
+
+theorem rank_le_of_forall_mem_rank_lt {o : Ordinal} :
+    (∀ y ∈ x, rank y < o) → rank x ≤ o :=
+  Quotient.inductionOn x fun _ h =>
+    PSet.rank_le_of_forall_mem_rank_lt fun y h' => h ⟦y⟧ h'
+
+theorem rank_le_of_subset (h : x ⊆ y) : rank x ≤ rank y :=
+  rank_le_of_forall_mem_rank_lt fun _ h₁ => rank_lt_of_mem (h h₁)
+
+@[simp] theorem rank_empty : rank ∅ = 0 := by
+  rw [← Ordinal.le_zero]; apply rank_le_of_forall_mem_rank_lt; simp
+
+@[simp] theorem rank_insert : rank (insert x y) = max (succ (rank x)) (rank y) := by
+  apply le_antisymm
+  · apply rank_le_of_forall_mem_rank_lt; simp
+    intro _ h; exact Or.inr (rank_lt_of_mem h)
+  · simp; constructor
```

Comment:
Using `simp` like this when it's not at the end of a proof is generally discouraged. If the set of applicable simp lemmas changes, the proof might suddenly and unexpectedly break.

You can use `simp?` to figure out what theorems are being called, and then use `simp only` with the theorems that are required. I think you'll find that in most cases, your `simp` can be replaced with a simple `rw` sequence.